src/com/google/enterprise/adaptor/DocIdPusher.java

```diff
@@ -184,6 +184,43 @@ public GroupPrincipal pushGroupDefinitions(
       throws InterruptedException;
 
   /**
+   * Blocking call to push group definitions to GSA ends in success or
+   * when provided error handler gives up.  Can take significant time
+   * if errors arise.
+   *
+   * <p>A group definition consists of a group being defined
+   * and members, which is a list of users and groups.
+   *
+   * <p>If you plan on using the return code, then the provided map should have
+   * a predictable iteration order, like {@link java.util.TreeMap}.
+   *
+   * <p>An incremental push augments or updates any existing group definitions
+   * for the groupSource on the GSA. Otherwise, the supplied group definitions
+   * will wholly replace all existing definitions for the groupSource on the
+   * GSA.
+   *
+   * <p>A groupSource identifies the source of the group definitions. If
+   * provided, it must be a string of the form {@code [a-zA-Z_][a-zA-Z0-9_-]*}.
+   * If groupSource is {@code null}, then the {@code feed.name} configuration
+   * property is used.
+   *
+   * <p>If handler is {@code null}, then a default error handler is used.
+   *
+   * @param defs map of group definitions
+   * @param caseSensitive when comparing Principals
+   * @param incremental if {@code true} incremental update is done, otherwise
+   *        full replacement is done.
+   * @param groupSource
+   * @param handler for dealing with errors pushing
+   * @return {@code null} on success, otherwise the first GroupPrincipal to fail
+   * @throws InterruptedException if interrupted and no definitions were sent
+   */
+  public GroupPrincipal pushGroupDefinitions(
```

Comment:
Adding this method to interface requires modification code(tests but still) for some adaptors (and thirdparty's as well). This doesn't looks like a minor changes and this was the main reason why my CL was rejected. In fact I am agree with this.

Last option that we discussed was the adding public method in abstract class and DocIdSender but not add it in interface. Yes, this require from AD connector to check and cast type to deal with new method but help us to avoid changing code in any adaptors.

Could you review this option?